arch/xtensa/src/esp32s3/esp32s3_psram.c

```diff
@@ -0,0 +1,559 @@
+#include "hardware/esp32s3_gpio_sigmap.h"
+
+/* EFUSE */
+
+#define ESP_ROM_EFUSE_FLASH_DEFAULT_SPI  (0)
+
+extern uint32_t esp_rom_efuse_get_flash_gpio_info(void);
+extern uint32_t esp_rom_efuse_get_flash_wp_gpio(void);
+
+/* Commands for PSRAM chip */
+
+#define PSRAM_READ                 0x03
+#define PSRAM_FAST_READ            0x0B
+#define PSRAM_FAST_READ_QUAD       0xEB
+#define PSRAM_WRITE                0x02
+#define PSRAM_QUAD_WRITE           0x38
+#define PSRAM_ENTER_QMODE          0x35
+#define PSRAM_EXIT_QMODE           0xF5
+#define PSRAM_RESET_EN             0x66
+#define PSRAM_RESET                0x99
+#define PSRAM_SET_BURST_LEN        0xC0
+#define PSRAM_DEVICE_ID            0x9F
+
+#define PSRAM_FAST_READ_DUMMY      4
+#define PSRAM_FAST_READ_QUAD_DUMMY 6
+
+/* ID */
+
+#define PSRAM_ID_KGD_M             0xff
+#define PSRAM_ID_KGD_S             8
+#define PSRAM_ID_KGD               0x5d
+#define PSRAM_ID_EID_M             0xff
+#define PSRAM_ID_EID_S             16
+
+/* Use the [7:5](bit7~bit5) of EID to distinguish the psram size:
+ *
+ *   BIT7  |  BIT6  |  BIT5  |  SIZE(MBIT)
+ *   -------------------------------------
+ *    0    |   0    |   0    |     16
+ *    0    |   0    |   1    |     32
+ *    0    |   1    |   0    |     64
+ */
+
+#define PSRAM_EID_SIZE_M           0x07
+#define PSRAM_EID_SIZE_S           5
+
+#define PSRAM_KGD(id)         (((id) >> PSRAM_ID_KGD_S) & PSRAM_ID_KGD_M)
+#define PSRAM_EID(id)         (((id) >> PSRAM_ID_EID_S) & PSRAM_ID_EID_M)
+#define PSRAM_SIZE_ID(id)     ((PSRAM_EID(id) >> PSRAM_EID_SIZE_S) & PSRAM_EID_SIZE_M)
+#define PSRAM_IS_VALID(id)    (PSRAM_KGD(id) == PSRAM_ID_KGD)
+
+#define PSRAM_IS_64MBIT_TRIAL(id) (PSRAM_EID(id) == 0x26)
+
+/* IO-pins for PSRAM.
+ * WARNING: PSRAM shares all but the CS and CLK pins with the flash, so
+ * these defines hardcode the flash pins as well, making this code
+ * incompatible with either a setup that has the flash on non-standard
+ * pins or ESP32s with built-in flash.
+ */
+
+#define FLASH_CLK_IO          SPI_CLK_GPIO_NUM
+#define FLASH_CS_IO           SPI_CS0_GPIO_NUM
+
+/* PSRAM clock and cs IO should be configured based on hardware design. */
+
+#define PSRAM_CLK_IO          CONFIG_ESP32S3_DEFAULT_PSRAM_CLK_IO /* Default value is 30 */
+#define PSRAM_CS_IO           CONFIG_ESP32S3_DEFAULT_PSRAM_CS_IO  /* Default value is 26 */
+#define PSRAM_SPIQ_SD0_IO     SPI_Q_GPIO_NUM
+#define PSRAM_SPID_SD1_IO     SPI_D_GPIO_NUM
+#define PSRAM_SPIWP_SD3_IO    SPI_WP_GPIO_NUM
+#define PSRAM_SPIHD_SD2_IO    SPI_HD_GPIO_NUM
+
+#define CS_PSRAM_SEL          SPI_MEM_CS1_DIS_M
+#define CS_FLASH_SEL          SPI_MEM_CS0_DIS_M
+
+#define SPI1_NUM              1
+#define SPI0_NUM              0
+
+typedef enum
+{
+  PSRAM_CMD_QPI,
+  PSRAM_CMD_SPI,
+} psram_cmd_mode_e;
+
+uint32_t g_psram_id;
+uint32_t g_psram_size;        /* physical psram size in bytes */
+
+static void config_psram_spi_phases(void);
+extern void esp_rom_spi_set_op_mode(int spi_num,
```

Comment:
can we group it together with
```
extern uint32_t esp_rom_efuse_get_flash_gpio_info(void);
extern uint32_t esp_rom_efuse_get_flash_wp_gpio(void);
```
?

Reply:
BTW I see this prototype in `arch/xtensa/src/esp32s3/rom/esp32s3_opi_flash.h`

Reply:
@pkarashchenko esp_rom_efuse_get_flash_gpio_info() and esp_rom_efuse_get_flash_wp_gpio() are generic for any kind of flash. The esp_rom_opiflash_pin_config() is only for Octa SPI Flash, so I cannot move it to the esp32s3_opi_flash.h file.

Reply:
I was writing about `esp_rom_spi_set_op_mode`, I thought that maybe `#include esp32s3_opi_flash.h` might be used instead of the prototype. The comment was not related to moving `esp_rom_efuse_...` APIs to `esp32s3_opi_flash.h`.

Reply:
There is a line at the top of the file `#include "rom/esp32s3_opi_flash.h"